File: CervejaDAO.php
<?php 
    include_once("conexao.php");
    include_once("Cerveja.php");

    class CervejaDAO{
        private $conexao;

        public function __construct(){
            $this->conexao = Conexao::getConexao();
        }

        public function inserir(Cerveja $cerveja){        
    $pstmt = $this->conexao->prepare(
        "INSERT INTO cerveja (
            nome, ibu, data_fabri, fabricante, tipo, local_degustado, 
            avaliacao, comentarios, img, pais_origem, teor
        ) VALUES (
            :nome, :ibu, :data_fabri, :fabricante, :tipo, :local_degustado,
            :avaliacao, :comentarios, :img, :pais_origem, :teor
        )"
    );

    $pstmt->bindValue(":nome", $cerveja->getNome());
    $pstmt->bindValue(":ibu", $cerveja->getIbu());
    $pstmt->bindValue(":data_fabri", $cerveja->getData());
    $pstmt->bindValue(":fabricante", $cerveja->getFabricante());
    $pstmt->bindValue(":tipo", $cerveja->getTipo());
    $pstmt->bindValue(":local_degustado", $cerveja->getLocal());
    $pstmt->bindValue(":avaliacao", $cerveja->getAvaliacao());
    $pstmt->bindValue(":comentarios", $cerveja->getComentarios());
    $pstmt->bindValue(":img", $cerveja->getImg());
    $pstmt->bindValue(":pais_origem", $cerveja->getPais_origem());
    $pstmt->bindValue(":teor", $cerveja->getTeor());

    $pstmt->execute();
}


       public function listar() {
    $vetor = [];
    $pstmt = $this->conexao->prepare("SELECT * FROM cerveja");
    $pstmt->execute();

    while ($linha = $pstmt->fetch(PDO::FETCH_ASSOC)) {
        $vetor[] = new Cerveja([
            "nome" => $linha["nome"],
            "ibu" => $linha["ibu"],
            "data_fabri" => $linha["data_fabri"],
            "fabricante" => $linha["fabricante"],
            "tipo" => $linha["tipo"],
            "teor" => $linha["teor"],
            "local_degustado" => $linha["local_degustado"],
            "avaliacao" => $linha["avaliacao"],
            "comentarios" => $linha["comentarios"],
            "img" => $linha["img"],
            "pais_origem" => $linha["pais_origem"]
        ]);
    }

    return $vetor;
}


        public function ListarPais(){
            $vetor = [];
            $pstmt = $this->conexao->prepare("SELECT * FROM cerveja ORDER BY pais DESC");
            $pstmt->execute();
            while($linha = $pstmt->fetch()){
                $vetor[] = new cerveja(
                $linha["nome"],
                $linha["ibu"],
                $linha["data_fabri"],
                $linha["fabricante"],
                $linha["tipo"],
                $linha["teor"],
                $linha["local_degustado"],
                $linha["avaliacao"],
                $linha["comentarios"],
                $linha["img"],
                $linha["pais_origem"],
                $linha["sugestao"],
                null);
            } 
            return $vetor;
        }

        public function listarNome(){
            $vetor = [];
            $pstmt = $this->conexao->prepare("SELECT * FROM cerveja ORDER BY nome DESC");
            $pstmt->execute();
            while($linha = $pstmt->fetch()){
                $vetor[] = new cerveja(
                $linha["nome"],
                $linha["ibu"],
                $linha["data_fabri"],
                $linha["fabricante"],
                $linha["tipo"],
                $linha["teor"],
                $linha["local_degustado"],
                $linha["avaliacao"],
                $linha["comentarios"],
                $linha["img"],
                $linha["pais_origem"],
                $linha["sugestao"],
                null);
            } 
            return $vetor;
        }

        public function listarNota(){
            $vetor = [];
            $pstmt = $this->conexao->prepare("SELECT * FROM cerveja ORDER BY avalicao DESC");
            $pstmt->execute();
            while($linha = $pstmt->fetch()){
                $vetor[] = new cerveja(
                $linha["nome"],
                $linha["ibu"],
                $linha["data_fabri"],
                $linha["fabricante"],
                $linha["tipo"],
                $linha["teor"],
                $linha["local_degustado"],
                $linha["avaliacao"],
                $linha["comentarios"],
                $linha["img"],
                $linha["pais_origem"],
                $linha["sugestao"],
                null);
            } 
            return $vetor;
        }

        public function listarEstilo($tipo){
            $vetor = [];
            $pstmt = $this->conexao->prepare("SELECT * FROM cerveja WHERE tipo=:tipo");
            $pstmt->bindValue(":tipo",$tipo);
            $pstmt->execute();
            while($linha = $pstmt->fetch()){
                $vetor[] = new cerveja(
                $linha["nome"],
                $linha["ibu"],
                $linha["data_fabri"],
                $linha["fabricante"],
                $linha["tipo"],
                $linha["teor"],
                $linha["local_degustado"],
                $linha["avaliacao"],
                $linha["comentarios"],
                $linha["img"],
                $linha["pais_origem"],
                $linha["sugestao"],
                null);
            } 
            return $vetor;
        }

        public function listarAvalicao($aval){
            $vetor = [];
            $pstmt = $this->conexao->prepare("SELECT * FROM cerveja WHERE avaliacao=:avaliacao");
            $pstmt->bindValue(":avaliacao",$aval);
            $pstmt->execute();
            while($linha = $pstmt->fetch()){
                $vetor[] = new cerveja(
                $linha["nome"],
                $linha["ibu"],
                $linha["data_fabri"],
                $linha["fabricante"],
                $linha["tipo"],
                $linha["teor"],
                $linha["local_degustado"],
                $linha["avaliacao"],
                $linha["comentarios"],
                $linha["img"],
                $linha["pais_origem"],
                $linha["sugestao"],
                null);
            } 
            return $vetor;
        }

        public function listarData($data){
            $vetor = [];
            $pstmt = $this->conexao->prepare("SELECT * FROM cerveja WHERE data_fabri=:data_fabri");
            $pstmt->bindValue(":data_fabri",$data);
            $pstmt->execute();
            while($linha = $pstmt->fetch()){
                $vetor[] = new cerveja(
                $linha["nome"],
                $linha["ibu"],
                $linha["data_fabri"],
                $linha["fabricante"],
                $linha["tipo"],
                $linha["teor"],
                $linha["local_degustado"],
                $linha["avaliacao"],
                $linha["comentarios"],
                $linha["img"],
                $linha["pais_origem"],
                $linha["sugestao"],
                null);
            } 
            return $vetor;
        }

        public function listarImgs() { 
        $lista = []; 
        $sql = "SELECT img FROM cerveja WHERE img IS NOT NULL"; $stmt = $this->conexao->query($sql);

        while($row = $stmt->fetch(PDO::FETCH_ASSOC)){
            $lista[] = $row['img'];
        }

        return $lista;
        } 

        public function getDegustacoesPorMes() {
            $vetor = [];
            
            $pstmt = $this->conexao->prepare(
                "SELECT 
                    MONTH(data_fabri) as mes_numero,
                    MONTHNAME(data_fabri) as mes_nome,
                    COUNT(*) as total_degustacoes
                FROM cerveja 
                GROUP BY MONTH(data_fabri), MONTHNAME(data_fabri)
                ORDER BY mes_numero"
            );
            
            $pstmt->execute();
            
            while($linha = $pstmt->fetch(PDO::FETCH_ASSOC)) {
                $vetor[] = [
                    'mes_numero' => $linha['mes_numero'],
                    'mes_nome' => $linha['mes_nome'],
                    'total_degustacoes' => $linha['total_degustacoes']
                ];
            }
            
            return $vetor;
        }

        public function listarMediaNotasPorTipo() {
            $vetor = [];
            
            $pstmt = $this->conexao->prepare(
                "SELECT tipo, AVG(avaliacao) as media_nota 
                FROM cerveja 
                GROUP BY tipo 
                ORDER BY media_nota DESC"
            );
            
            $pstmt->execute();
            
            while($linha = $pstmt->fetch(PDO::FETCH_ASSOC)) {
                $vetor[] = [
                    'tipo' => $linha['tipo'],
                    'media_nota' => $linha['media_nota']
                ];
            }
            
            return $vetor;
        }
    }
?>

File: CervejaDAO.java
import java.sql.Connection;
import java.sql.Date;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.LinkedList;

public class CervejaDAO {
private Connection conexao;
	
	public CervejaDAO() {
		this.conexao = Conexao.getConexao();
	}
	
public boolean adicionar(Cerveja cerv) {
		
	String sql = "INSERT INTO cerveja (nome, tipo, teor_alcolico, IBU, pais_origem, fabricante, data_degustacao, avaliacao, comentarios, img,iduser) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?,?)";
		try {
			PreparedStatement stmt = (PreparedStatement) this.conexao.prepareStatement(sql);
			
			stmt.setString(1, cerv.getNome());
            stmt.setString(2, cerv.getTipo());
            stmt.setInt(3, cerv.getTeor_alcolico());
            stmt.setInt(4, cerv.getIBU());
            stmt.setString(5, cerv.getPais_origem());
            stmt.setString(6, cerv.getFabricante());
            stmt.setDate(7, cerv.getData_Degustacao());
            stmt.setDouble(8, cerv.getAvaliacao());
            stmt.setString(9, cerv.getComentarios());
            stmt.setString(10, cerv.getImg());
            stmt.setInt(11, cerv.getIduser());
			stmt.execute();
			stmt.close();
			return true;
		}catch(SQLException e) {
			System.out.println(e);
			return false;
		}
	}

public LinkedList<Cerveja> listar(int id) {
	String sql = "SELECT * FROM cerveja WHERE iduser = ?";
	
	try {
		LinkedList<Cerveja> cervejas = new LinkedList<Cerveja>();
		PreparedStatement stmt = this.conexao.prepareStatement(sql);
		stmt.setInt(1, id);
		ResultSet rs = stmt.executeQuery();

		while (rs.next()) {
			
			Cerveja c = new Cerveja(rs.getString("nome"),
	                rs.getString("tipo"),
	                rs.getInt("teor_alcolico"),
	                rs.getInt("IBU"),
	                rs.getString("pais_origem"),
	                rs.getString("fabricante"),
	                rs.getDate("data_degustacao"),
	                rs.getDouble("avaliacao"),
	                rs.getString("comentarios"),
	                rs.getString("img"));
			c.setId(rs.getInt("id"));				

			cervejas.add(c);
		}
		rs.close();
		stmt.close();
		return cervejas;
	} catch (SQLException e) {
		throw new RuntimeException(e);
	}
}

public LinkedList<Cerveja> listarPorNota(int id) {
	String sql = "SELECT * FROM cerveja WHERE iduser = ? ORDER BY avaliacao DESC";
	
	try {
		LinkedList<Cerveja> cervejas = new LinkedList<Cerveja>();
		PreparedStatement stmt = this.conexao.prepareStatement(sql);
		stmt.setInt(1, id);
		ResultSet rs = stmt.executeQuery();

		while (rs.next()) {
			
			Cerveja c = new Cerveja(rs.getString("nome"),
	                rs.getString("tipo"),
	                rs.getInt("teor_alcolico"),
	                rs.getInt("IBU"),
	                rs.getString("pais_origem"),
	                rs.getString("fabricante"),
	                rs.getDate("data_degustacao"),
	                rs.getDouble("avaliacao"),
	                rs.getString("comentarios"),
	                rs.getString("img"));
			c.setId(rs.getInt("id"));				

			cervejas.add(c);
		}
		rs.close();
		stmt.close();
		return cervejas;
	} catch (SQLException e) {
		throw new RuntimeException(e);
	}
}



public LinkedList<Object[]> listarDegustacaoMensal(int id) {
    String sql = "SELECT TO_CHAR(data_degustacao, 'MM/YYYY') as mes_ano, COUNT(*) as quantidade " +
                 "FROM cerveja WHERE iduser = ? GROUP BY mes_ano ORDER BY mes_ano DESC";
    
    try {
        LinkedList<Object[]> resultado = new LinkedList<>();
        PreparedStatement stmt = this.conexao.prepareStatement(sql);
        stmt.setInt(1, id);
        ResultSet rs = stmt.executeQuery();

        while (rs.next()) {
            resultado.add(new Object[]{
                rs.getString("mes_ano"),
                rs.getInt("quantidade")
            });
        }
        rs.close();
        stmt.close();
        return resultado;
    } catch (SQLException e) {
        System.out.print(e);
    }
	return null;
}

public LinkedList<Cerveja> listarFiltroDate(int id,Date data) {
	String sql = "SELECT * FROM cerveja WHERE iduser = ? AND  data_degustacao = ?";
	
	try {
		LinkedList<Cerveja> cervejas = new LinkedList<Cerveja>();
		PreparedStatement stmt = this.conexao.prepareStatement(sql);
		stmt.setInt(1, id);
		stmt.setDate(2, data);
		ResultSet rs = stmt.executeQuery();

		while (rs.next()) {
			
			Cerveja c = new Cerveja(rs.getString("nome"),
	                rs.getString("tipo"),
	                rs.getInt("teor_alcolico"),
	                rs.getInt("IBU"),
	                rs.getString("pais_origem"),
	                rs.getString("fabricante"),
	                rs.getDate("data_degustacao"),
	                rs.getDouble("avaliacao"),
	                rs.getString("comentarios"),
	                rs.getString("img"));
			c.setId(rs.getInt("id"));				

			cervejas.add(c);
		}
		rs.close();
		stmt.close();
		return cervejas;
	} catch (SQLException e) {
		throw new RuntimeException(e);
	}
}

public LinkedList<Cerveja> listarFiltroNota(int id,double avaliacao) {
	String sql = "SELECT * FROM cerveja WHERE iduser = ? AND  avaliacao = ?";
	
	try {
		LinkedList<Cerveja> cervejas = new LinkedList<Cerveja>();
		PreparedStatement stmt = this.conexao.prepareStatement(sql);
		stmt.setInt(1, id);
		stmt.setDouble(2, avaliacao);
		ResultSet rs = stmt.executeQuery();

		while (rs.next()) {
			
			Cerveja c = new Cerveja(rs.getString("nome"),
	                rs.getString("tipo"),
	                rs.getInt("teor_alcolico"),
	                rs.getInt("IBU"),
	                rs.getString("pais_origem"),
	                rs.getString("fabricante"),
	                rs.getDate("data_degustacao"),
	                rs.getDouble("avaliacao"),
	                rs.getString("comentarios"),
	                rs.getString("img"));
			c.setId(rs.getInt("id"));				

			cervejas.add(c);
		}
		rs.close();
		stmt.close();
		return cervejas;
	} catch (SQLException e) {
		throw new RuntimeException(e);
	}
}

public Cerveja buscarPorId(int id) {
    String sql = "SELECT * FROM cerveja WHERE id = ?";
    
    try {
        PreparedStatement stmt = this.conexao.prepareStatement(sql);
        stmt.setInt(1, id);
        ResultSet rs = stmt.executeQuery();

        if (rs.next()) {
            return new Cerveja(
                rs.getString("nome"),
                rs.getString("tipo"),
                rs.getInt("teor_alcolico"),
                rs.getInt("IBU"),
                rs.getString("pais_origem"),
                rs.getString("fabricante"),
                rs.getDate("data_degustacao"),
                rs.getDouble("avaliacao"),
                rs.getString("comentarios"),
                rs.getString("img")
            );
        }
        rs.close();
        stmt.close();
        return null;
    } catch (SQLException e) {
        throw new RuntimeException(e);
    }
}

public boolean editar(int id, Cerveja cerv) {
    String sql = "UPDATE cerveja SET nome = ?, tipo = ?, teor_alcolico = ?, IBU = ?, pais_origem = ?,"
    		+ " fabricante = ?, data_degustacao = ?, avaliacao = ?, comentarios = ?, img = ? WHERE id = ?";
    
    try {
        PreparedStatement stmt = this.conexao.prepareStatement(sql);
        
        stmt.setString(1, cerv.getNome());
        stmt.setString(2, cerv.getTipo());
        stmt.setInt(3, cerv.getTeor_alcolico());
        stmt.setInt(4, cerv.getIBU());
        stmt.setString(5, cerv.getPais_origem());
        stmt.setString(6, cerv.getFabricante());
        stmt.setDate(7, cerv.getData_Degustacao());
        stmt.setDouble(8, cerv.getAvaliacao());
        stmt.setString(9, cerv.getComentarios());
        stmt.setString(10, cerv.getImg());
        stmt.setInt(11, id);
        
        int rowsAffected = stmt.executeUpdate();
        stmt.close();
        return rowsAffected > 0;
    } catch(SQLException e) {
        return false;
    }
}


public LinkedList<Object[]> listarMediaPorTipo(int id) {
    String sql = "SELECT tipo, AVG(avaliacao) as media, COUNT(*) as quantidade " +
                 "FROM cerveja WHERE iduser = ? GROUP BY tipo ORDER BY media DESC";
    
    try {
        LinkedList<Object[]> resultado = new LinkedList<>();
        PreparedStatement stmt = this.conexao.prepareStatement(sql);
        stmt.setInt(1, id);
        ResultSet rs = stmt.executeQuery();

        while (rs.next()) {
            resultado.add(new Object[]{
                rs.getString("tipo"),
                rs.getDouble("media"),
                rs.getInt("quantidade")
            });
        }
        rs.close();
        stmt.close();
        return resultado;
    } catch (SQLException e) {
        throw new RuntimeException(e);
    }
}




public LinkedList<Cerveja> listarPorNome(int id) {
    String sql = "SELECT * FROM cerveja WHERE iduser = ? ORDER BY nome ASC";
    
    try {
        LinkedList<Cerveja> cervejas = new LinkedList<Cerveja>();
        PreparedStatement stmt = this.conexao.prepareStatement(sql);
        stmt.setInt(1, id);
        ResultSet rs = stmt.executeQuery();

        while (rs.next()) {
            Cerveja c = new Cerveja(
                rs.getString("nome"),
                rs.getString("tipo"),
                rs.getInt("teor_alcolico"),
                rs.getInt("IBU"),
                rs.getString("pais_origem"),
                rs.getString("fabricante"),
                rs.getDate("data_degustacao"),
                rs.getDouble("avaliacao"),
                rs.getString("comentarios"),
                rs.getString("img")
            );
            c.setId(rs.getInt("id"));
            cervejas.add(c);
        }
        rs.close();
        stmt.close();
        return cervejas;
    } catch (SQLException e) {
        throw new RuntimeException(e);
    }
}

public LinkedList<Cerveja> listarPorPais(int id) {
    String sql = "SELECT * FROM cerveja WHERE iduser = ? ORDER BY pais_origem ASC";
    
    try {
        LinkedList<Cerveja> cervejas = new LinkedList<Cerveja>();
        PreparedStatement stmt = this.conexao.prepareStatement(sql);
        stmt.setInt(1, id);
        ResultSet rs = stmt.executeQuery();

        while (rs.next()) {
            Cerveja c = new Cerveja(
                rs.getString("nome"),
                rs.getString("tipo"),
                rs.getInt("teor_alcolico"),
                rs.getInt("IBU"),
                rs.getString("pais_origem"),
                rs.getString("fabricante"),
                rs.getDate("data_degustacao"),
                rs.getDouble("avaliacao"),
                rs.getString("comentarios"),
                rs.getString("img")
            );
            c.setId(rs.getInt("id"));
            cervejas.add(c);
        }
        rs.close();
        stmt.close();
        return cervejas;
    } catch (SQLException e) {
        throw new RuntimeException(e);
    }
}

public LinkedList<Cerveja> listarPorData(int id) {
    String sql = "SELECT * FROM cerveja WHERE iduser = ? ORDER BY data_degustacao DESC";
    
    try {
        LinkedList<Cerveja> cervejas = new LinkedList<Cerveja>();
        PreparedStatement stmt = this.conexao.prepareStatement(sql);
        stmt.setInt(1, id);
        ResultSet rs = stmt.executeQuery();

        while (rs.next()) {
            Cerveja c = new Cerveja(
                rs.getString("nome"),
                rs.getString("tipo"),
                rs.getInt("teor_alcolico"),
                rs.getInt("IBU"),
                rs.getString("pais_origem"),
                rs.getString("fabricante"),
                rs.getDate("data_degustacao"),
                rs.getDouble("avaliacao"),
                rs.getString("comentarios"),
                rs.getString("img")
            );
            c.setId(rs.getInt("id"));
            cervejas.add(c);
        }
        rs.close();
        stmt.close();
        return cervejas;
    } catch (SQLException e) {
        throw new RuntimeException(e);
    }
}
}
